### PR TITLE
storage: consider source's sinces

### LINKED ISSUE #19151: Materialized views are created with invalid sinces
### What version of Materialize are you using?

main

### What is the issue?

Materialized views are created with `since`s that are before the `since`s of their sources.

```sql
materialize=> create table t (a int);
explain timestamp for select * from t;
create materialized view mv as select * from t;
explain timestamp for select * from mv;
CREATE TABLE
                                 Timestamp
---------------------------------------------------------------------------
                 query timestamp: 1683555414222 (2023-05-08 14:16:54.222) +
           oracle read timestamp: 1683555414222 (2023-05-08 14:16:54.222) +
 largest not in advance of upper: 1683555414222 (2023-05-08 14:16:54.222) +
                           upper:[1683555414223 (2023-05-08 14:16:54.223)]+
                           since:[1683555414222 (2023-05-08 14:16:54.222)]+
         can respond immediately: true                                    +
                        timeline: Some(EpochMilliseconds)                 +
                                                                          +
 source materialize.public.t (u1, storage):                               +
                   read frontier:[1683555414222 (2023-05-08 14:16:54.222)]+
                  write frontier:[1683555414223 (2023-05-08 14:16:54.223)]+

(1 row)

CREATE MATERIALIZED VIEW
                                 Timestamp
---------------------------------------------------------------------------
                 query timestamp: 1683555414457 (2023-05-08 14:16:54.457) +
           oracle read timestamp: 1683555414457 (2023-05-08 14:16:54.457) +
 largest not in advance of upper: 1683555414221 (2023-05-08 14:16:54.221) +
                           upper:[1683555414222 (2023-05-08 14:16:54.222)]+
                           since:[1683555413000 (2023-05-08 14:16:53.000)]+
         can respond immediately: false                                   +
                        timeline: Some(EpochMilliseconds)                 +
                                                                          +
 source materialize.public.mv (u2, storage):                              +
                   read frontier:[1683555413000 (2023-05-08 14:16:53.000)]+
                  write frontier:[1683555414222 (2023-05-08 14:16:54.222)]+

(1 row)
```

Note how the `since` of the MV (1683555413000) is before the `since` of the table it reads from (1683555414222). That is a bug, because the MV can't possibly be valid for reading when the thing it reads from is not.

### Initial Diagnosis

When creating an MV, adapter tries to ensure that the MV's `since` is >= the `since`s of the MV's sources by setting the `CollectionDescription::since` field in the `create_collections` call to the storage controller:
https://github.com/MaterializeInc/materialize/blob/e78dfcdd34c621c7f586409557c715f6dce1ad6e/src/adapter/src/coord/sequencer/inner.rs#L1267

Further up in the code, the value passed is specified to be the minimum valid read time on the MV’s sources, which is equal to the table's `since` in the example above.

The issue is that `CollectionDescription::since` is never looked at by the storage controller. So what the adapter specifies there is ignored and we end up with an invalid `since`. The commit that made the storage controller stop taking the specified since into account appears to be https://github.com/MaterializeInc/materialize/pull/17284/commits/e2a25499fbce56b858e57c46d9179fea38ffea96.

### PR DESCRIPTION
**Pull request by sploiselle**
fixes #19151

@MaterializeInc/qa Do you all have a sense of how we can test this fix? Comparing these values isn't to easy, though I've validated the fix locally:

```
materialize=> create table t (a int);
explain timestamp for select * from t;
create materialized view mv as select * from t;
explain timestamp for select * from mv;
CREATE TABLE
                                 Timestamp                                 
---------------------------------------------------------------------------
                 query timestamp: 1683559211046 (2023-05-08 15:20:11.046) +
           oracle read timestamp: 1683559211046 (2023-05-08 15:20:11.046) +
 largest not in advance of upper: 1683559211046 (2023-05-08 15:20:11.046) +
                           upper:[1683559211047 (2023-05-08 15:20:11.047)]+
                           since:[1683559211046 (2023-05-08 15:20:11.046)]+
         can respond immediately: true                                    +
                        timeline: Some(EpochMilliseconds)                 +
                                                                          +
 source materialize.public.t (u3, storage):                               +
                   read frontier:[1683559211046 (2023-05-08 15:20:11.046)]+
                  write frontier:[1683559211047 (2023-05-08 15:20:11.047)]+
 
(1 row)

CREATE MATERIALIZED VIEW
                                 Timestamp                                 
---------------------------------------------------------------------------
                 query timestamp: 1683559211227 (2023-05-08 15:20:11.227) +
           oracle read timestamp: 1683559211227 (2023-05-08 15:20:11.227) +
 largest not in advance of upper: 1683559211045 (2023-05-08 15:20:11.045) +
                           upper:[1683559211046 (2023-05-08 15:20:11.046)]+
                           since:[1683559211046 (2023-05-08 15:20:11.046)]+
         can respond immediately: false                                   +
                        timeline: Some(EpochMilliseconds)                 +
                                                                          +
 source materialize.public.mv (u4, storage):                              +
                   read frontier:[1683559211046 (2023-05-08 15:20:11.046)]+
                  write frontier:[1683559211046 (2023-05-08 15:20:11.046)]+
 
(1 row)

materialize=> SELECT 1683559211046 <= 1683559211046;
 ?column? 
----------
 t
(1 row)
```

The issue before was that the since of the view `mv` could be less than the since of `t`.

### Motivation

 This PR fixes a recognized bug. Fixes #19151

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/storage-client/src/controller/remap_migration.rs [hunk 1]
```diff
@@ -1,125 +0,0 @@
-// Copyright Materialize, Inc. and contributors. All rights reserved.
```
Comment (aljoscha) [inline src/storage-client/src/controller/remap_migration.rs:1]: 🎉

**Review by aljoscha — APPROVED** [review]
Changes look very good! I had one proposal about tweaking how we chose the `since`.

File: src/storage-client/src/controller.rs [hunk 2]
```diff
@@ -2383,7 +2385,10 @@ where
                 .await
                 .expect("invalid persist usage");
 
-            let since = handle.since().clone();
+            let since = collection_desc
```
[the remark this answers is not in the document]
Comment (sploiselle) [reply src/storage-client/src/controller.rs]: Oh good idea! Done

**Comment by def-** [comment]
I think with the current tooling not so easy to test automatically. A Python test would have to parse the outputs of `explain timestamp`, store them in a ExplainTimestamp object and compare them then. Would that help?

**Comment by def-** [comment]
I have added a Python-based cluster test for this (`bin/mzcompose --find cluster down && bin/mzcompose --release --find cluster run test-mv-source-sink`), but have seen a sporadic failure:
```
AssertionError: "since" timestamp of mv (1683631436137) is less than "since" timestamp of its source table (1683631437000)
```
@sploiselle Is this expected or should this work? My assumption was that no sleeps should be required to get this to match.

**Comment by aljoscha** [comment]
@sploiselle I just merged https://github.com/MaterializeInc/materialize/pull/19205, which should unblock this 🎉

**Comment by sploiselle** [comment]
@def- tysm for this test--a huge help. we had some issues on the db side but have resolved them and the test now reliably passes. I've amended your description and pushed everything up.

**Pull request merged**